src/reducers.js

```diff
@@ -0,0 +1,182 @@
+import { predefinedGames } from './utils/predefined-games';
+import bots from './bots';
+import { interactiveCellTypes } from './utils/constants';
+import { CLICK_CELL, CHANGE_GAME_INDEX, RESET_GAME, CHANGE_BOT, BOT_RUNNING } from './actions';
+import { combineReducers } from 'redux';
+
+// Initial game index to load.
+const INITIAL_GAME_INDEX = 0;
+
+/**
+   Helper function that returns the next cell type when a cell with a type has been clicked.
+   @method _nextCellType
+   @param {String} type Cell type.
+   @return {String}
+*/
+function _nextCellType(type) {
+    const index = interactiveCellTypes.indexOf(type);
+
+    if (index < 0) {
+        return 'empty';
+    }
+
+    return interactiveCellTypes[(index + 1) % interactiveCellTypes.length];
+}
+
+/**
+   Game state reducer.
+*/
+function game(state = predefinedGames[INITIAL_GAME_INDEX], action) {
+
+    // Game index and reset.  Return a predefined game.
+    if (action.type === CHANGE_GAME_INDEX || action.type === RESET_GAME) {
+        return predefinedGames[action.index];
+    }
+
+    // If its any other action we don't know about, just return state back.
+    if (action.type !== CLICK_CELL) {
+        return state;
+    }
+
+    // Out of bounds error. Just return same board.
+    if (action.row < 1 ||
+        action.row >= state.board.length ||
+        action.column < 1 ||
+        action.column >= state.board[0].length ||
+        state.board[action.row][action.column].type === 'coral') {
+        return state;
+    }
+
+    // Create new immutable board.
+    const board = state.board.map(row => {
+        return row.map(cell => Object.assign({}, cell));
+    });
+
+    // Update cell type based on what was clicked.
+    const cell = board[action.row][action.column];
+    cell.type = _nextCellType(cell.type);
+
+    const rowConstraintCell = board[action.row][0];
+    const colConstraintCell = board[0][action.column];
+
+    // Update row constraints state.
+    const row = board[action.row];
+    const column = board.map(row => row[action.column]);
+
+    if (!row.some(cell => cell.type === 'empty')) {
+        let numClownfishs = 0;
+
+        row.forEach(cell => {
+            if (cell.type === 'clownfish') {
+                numClownfishs++;
+            }
+        });
```

Comment:
A pattern I really like for counting like this in JS is: 

```
const numClownfishs = row.filter(cell => cell.type === 'clownfish').length
```

Reply:
Ha yep.  Just forgot about this one =)  Functional patterns are the best.